File: src/Solutions/Euler088.hs
module Solutions.Euler088 where
import Utils.List(uniques)

-- productSum k \in [k, 2k]
-- since each ai >= 1 and example product sum is a1 = k, a2 = 2, a3 = .. = a_k = 1

check :: Int -> Int -> Int -> Int -> Bool
check remLen remProd remSum maxN
    | remLen == 0 = remProd == 1 && remSum == 0
    | remProd < 1 = False
    | remProd == 1 = remSum == remLen -- fill remaining holes with 1
    | remProd > 1 = any (\x -> check (remLen - 1) (remProd `div` x) (remSum - x) x)
                       $ filter ((==0) . mod remProd) [2..maxN]

iter :: Int -> Int -> Int
iter n k = if check k n n k then n else iter (n + 1) k

productSum :: Int -> Int
productSum k = iter k k

solution :: Int -> Int
solution n = sum $ uniques $ map productSum [2..n]

euler088 :: IO String
euler088 = return $ show $ solution 12000

File: rust/src/solutions/euler088.rs
fn check(rem_len: i64, rem_prod: i64, rem_sum: i64, max_n: i64) -> bool {
    if rem_len == 0 {
        rem_prod == 1 && rem_sum == 0
    } else {
        match rem_prod {
            1 => rem_sum == rem_len,
            _ if rem_prod < 1 => false,
            _ => (2..=max_n)
                .filter(|&n| rem_prod % n == 0)
                .any(|n| check(rem_len - 1, rem_prod / n, rem_sum - n, n)),
        }
    }
}

fn iter(n: i64, k: i64) -> i64 {
    if check(k, n, n, k) {
        n
    } else {
        iter(n + 1, k)
    }
}

fn product_sum(k: i64) -> i64 {
    iter(k, k)
}

pub fn euler088() -> String {
    let mut psums = (2..=12000).map(product_sum).collect::<Vec<i64>>();
    psums.sort();
    psums.dedup();
    psums.iter().sum::<i64>().to_string()
}
